Optionals v1

joshring opened the pull request:
Large scale rework of the optionals section aiming to be easier to grasp for beginners who have not done functional languages previously. I fully expect there will be parts which need tweaking

## This is the thought process

- Giving a summary at the start to get you started
- Examples follow next, leaning more heavily on being easy to run and being much earlier in the text, encouraging exploration
- A lot of the same ideas that were there before, are there now also but have been re-organised or fleshed out more. 
- Aiming to prioritise what the reader needs to know first because this is one of the more challenging aspects of the language for newer developers.

joshring committed:
hotfix of indexing

joshring committed:
Revert "hotfix of indexing"

This reverts commit 352b5947bbf74a88bc62efc942c867d6b8e7f6cc.

joshring committed:
Merge remote-tracking branch 'origin/main'

joshring committed:
Merge remote-tracking branch 'origin/main'

joshring committed:
Merge remote-tracking branch 'origin/main'

joshring committed:
Merge remote-tracking branch 'origin/main'

joshring committed:
large scale rework of the optionals section aiming to be easier to grasp for beginners who have not done functional languages previously.

lerno commented on src/content/docs/references/docs/optionals.md:
```diff
@@ -5,187 +5,272 @@ sidebar:
-4. A function called with an Optional is only invoked if it has an actual value.
- 
-## What is an Optional?
+A common example using optionals is for a function that can return a value or it can fail. We call the reason for the failure a `fault` and the successful outcome a `result`. 
```
Can we further stress that it's just that a function might not be able to provide a value rather than that something needs to be an error?

lerno commented on src/content/docs/references/docs/optionals.md:
```diff
@@ -5,187 +5,272 @@ sidebar:
 
-    int! a = ...
-    fn double bar(int x) { ... }
+Optionals are a tagged union of either the `result` **or** `fault`, either of which can be unpacked in C3.
```
Replace "are a" with "act like a", because there are implications it was a tagged union, like for example one might think that an optional could be stored in a struct or be the parameter type of a call.

lerno commented on src/content/docs/references/docs/optionals.md:
````diff
@@ -5,187 +5,272 @@ sidebar:
-    int! z = baz(a, a);  // same as z = bar(1, 1)
+#### Run code if we find a `fault` in the Optional: 
+```c3
+// Detect `fault` inside "optional_value", handle inside scope
````
This is hard. Typically for Optionals we talk about it being "empty" or "contains a value". If we use the word "has a fault", then this is weird: `anyfault! x;` which is quite fine in C3. I think of the C3 optional having 3 things really:

1. It can contain a value, optionally
2. It can be "empty", in which case there isn't a value.
3. IF it is empty it also contains a *reason* why it's empty which is the *fault* (although I have certainly tried finding alternative names to *fault* that would give less association to an error)

These terms are established:
1. When we simply use `if (catch foo)` then we "test it if it's empty" or "test for Empty". 
2. If we do `if (try foo())` we "test if it has a value" or "test for Value".
3. When we use `if (try foo = bar())` we "unwrap the optional"

However, when we have `if (catch f = foo)`, then no terms are established. Maybe it could be like "unwrap the Empty's fault" "get the underlying fault" "get the reason for Empty"

joshring replied:
terminology agreed:

> testing if a value is missing `if (catch optional_value)`
testing if a value is present  `if (try optional_value)`
unwrapping a value with `if (try value = optional_value)`
retrieve a fault with `if (catch reason = optional_value)`

joshring committed:
good clarifications from comments

joshring committed:
reworked to use new terminology

testing if a value is missing if (catch optional_value)
testing if a value is present  if (try optional_value)
unwrapping a value with if (try value = optional_value)
retrieve a fault with if (catch reason = optional_value)

reworking structure and reducing duplication

joshring committed:
sharpening wording

joshring committed:
tweaks

joshring committed:
more complete examples

joshring committed:
cleanup continued

joshring committed:
updating examples

lerno commented on src/content/docs/references/docs/optionals.md:
```diff
@@ -5,571 +5,736 @@ sidebar:
-4. A function called with an Optional is only invoked if it has an actual value.
- 
-## What is an Optional?
+A commonly Optionals are used for a function that can return a value that is *present* or return a value that is *missing*, along with a *reason* why it was missing. We call the intended outcome a `value`, when it could not return a `value` we call that *reason* the `fault`.
```
Let's go all in and call it "the *Excuse*"

joshring committed:
all in on excuses

joshring committed:
going more all in on Excuse
